src/main/java/org/openrewrite/java/template/processor/RefasterTemplateProcessor.java

```diff
@@ -124,528 +126,536 @@ public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment
-
-                SortedSet<String> visitMethods = new TreeSet<>();
-                beforeTemplates.entrySet().stream()
-                        .collect(groupingBy(
-                                entry -> getType(entry.getValue().method),
-                                toMap(Map.Entry::getKey, Map.Entry::getValue)))
-                        .forEach((type, templates) -> {
-                            for (String lstType : LST_TYPE_MAP.get(type)) {
-                                visitMethods.add(generateVisitMethod(templates, after, descriptor, lstType));
-                            }
-                        });
-                visitMethods.forEach(visitor::append);
-                visitor.append("        }");
-                return visitor.toString();
             }
-
-            private String generateVisitMethod(Map<String, TemplateDescriptor> beforeTemplates, String after, RuleDescriptor descriptor, String lstType) {
-                StringBuilder visitMethod = new StringBuilder();
-                String methodSuffix = lstType.startsWith("J.") ? lstType.substring(2) : lstType;
-                visitMethod.append("            @Override\n");
-                visitMethod.append("            public J visit").append(methodSuffix).append("(").append(lstType).append(" elem, ExecutionContext ctx) {\n");
-                if (lstType.equals("Statement")) {
-                    visitMethod.append("                if (elem instanceof J.Block) {\n");
-                    visitMethod.append("                    // FIXME workaround\n");
-                    visitMethod.append("                    return elem;\n");
-                    visitMethod.append("                }\n");
+            if (after != null) {
+                visitor.append("            final JavaTemplate ")
+                        .append(after)
+                        .append(" = ")
+                        .append(descriptor.afterTemplate.toJavaTemplateBuilder())
+                        .append("\n                    .build();\n");
+            }
+            visitor.append("\n");
+
+            // Determine which visitMethods we should generate
+            Map<String, Map<String, TemplateDescriptor>> templatesByLstType = new TreeMap<>();
+            for (Map.Entry<String, TemplateDescriptor> entry : beforeTemplates.entrySet()) {
+                Set<Class<? extends JCTree>> types = entry.getValue().getTypes();
+                for (Class<? extends JCTree> type : types) {
+                    for (String lstType : LST_TYPE_MAP.get(type)) {
+                        templatesByLstType.computeIfAbsent(lstType, k -> new TreeMap<>())
+                                .put(entry.getKey(), entry.getValue());
+                    }
                 }
+            }
+            templatesByLstType.forEach((lstType, typeBeforeTemplates) ->
+                    visitor.append(generateVisitMethod(typeBeforeTemplates, after, descriptor, lstType)));
```

Comment:
This is the heart of the changes here: previously we used `groupingBy` with a faulty `getType` method, that instead should support multiple types when it handles `Refaster.anyOf` correctly.